build: more type information for docs

DevVersion opened the pull request:
* Upgrades dgeni and dgeni-packages to the latest available version (including TypeScript support; and newer TypeScript version for AST)
* More warnings if type information is missing on public TypeScript symbol
* Adds missing explicit type declarations to all components
* Types are now being displayed for accessors and properties (this was broken before somehow)
* Rewrote the Dgeni setup in TypeScript for type safety and better integration into the gulp workflow.

**Note**: Most of the things inside of the existing Dgeni setup is the same as before and I didn't look into other improvements there (e.g logic simplification) because the PR is big enough already.

crisbeto commented on src/lib/form-field/form-field.ts:
```diff
@@ -94,7 +94,7 @@ export class MdFormField implements AfterViewInit, AfterContentInit, AfterConten
   /** @deprecated Use `color` instead. */
   @Input()
-  get dividerColor() { return this.color; }
+  get dividerColor(): 'primary' | 'accent' | 'warn' { return this.color; }
```
We have the `ThemePalette` type for this already.

DevVersion replied:
Yeah I wanted to use that as well, but it also includes `undefined` as a possible value, which is not valid here.

crisbeto commented on tools/dgeni/common/decorators.ts:
```diff
@@ -0,0 +1,113 @@
+import {PropertyMemberDoc} from 'dgeni-packages/typescript/api-doc-types/PropertyMemberDoc';
+import {MemberDoc} from 'dgeni-packages/typescript/api-doc-types/MemberDoc';
+
+const SELECTOR_BLACKLIST  = new Set([
```
Some inconsistent spacing here.

crisbeto commented on tools/dgeni/processors/categorizer.ts:
```diff
@@ -0,0 +1,129 @@
+import {ClassExportDoc} from 'dgeni-packages/typescript/api-doc-types/ClassExportDoc';
+import {PropertyMemberDoc} from 'dgeni-packages/typescript/api-doc-types/PropertyMemberDoc';
+import {
+  NormalizedMethodMemberDoc, normalizeMethodParameters
```
The `normalizeMethodParameters` should be on a separate line.

DevVersion commented:
@crisbeto Addressed your feedback and rebased on top of the [sort members](https://github.com/angular/material2/commit/9f2fd477f24cda4e634209510162d171572ccaa0) commit. 

Regarding the enforcement: Good point, I have a TSLint rule in mind. We can base those rules on top of the existing rules like `typedef` (will be something as a follow-up)

DevVersion commented:
@jelbourn Rebased this one. Please have another look.

jelbourn commented on tools/dgeni/common/decorators.ts:
```diff
@@ -0,0 +1,113 @@
+
+export function isProperty(doc: MemberDoc) {
+  if (doc instanceof PropertyMemberDoc ||
+      // The latest Dgeni version no longer treats getters or setters as properties. In theory
```
"In theory those" -> "From a user perspective, these"

jelbourn commented on tools/dgeni/common/dgeni-accessors-parse.ts:
```diff
@@ -0,0 +1,54 @@
+import {CategorizedPropertyMemberDoc} from '../processors/categorizer';
+
+/**
+ * Accessors are not parsed properly within the latest dgeni-packages package.
```
Is there an issue to link for this? What about the accessors isn't parsed correctly?

DevVersion replied:
I only reported this on Slack if I recall. Created an issue for it now: https://github.com/angular/dgeni-packages/issues/246

jelbourn commented on tools/dgeni/common/dgeni-accessors-parse.ts:
```diff
@@ -0,0 +1,54 @@
+        member.kind === SyntaxKind.SetAccessor)
+      .filter((member: any) => member.name && (member.name as any).text === propertyDoc.name)
+      .forEach((member: GetAccessorDeclaration|SetAccessorDeclaration) => {
+        if (member.type) {
```
Add comment that explains how you extract the type

jelbourn commented on tools/dgeni/common/dgeni-accessors-parse.ts:
```diff
@@ -0,0 +1,54 @@
+    classDeclaration.members
+      .filter((member: any) => member.kind === SyntaxKind.GetAccessor ||
+        member.kind === SyntaxKind.SetAccessor)
+      .filter((member: any) => member.name && (member.name as any).text === propertyDoc.name)
```
If `member` is `any`, why do you need to use `as any` on `member.name`?

jelbourn commented on tools/dgeni/common/dgeni-accessors-parse.ts:
```diff
@@ -0,0 +1,54 @@
+        // If the current property document doesn't have any description set and the current
+        // TypeScript has a symbol from the TypeChecker, we can try to find a documentation
+        // comment on that symbol.
+        if (!propertyDoc.description && (member as any).symbol) {
```
If you need the `as any` here, is `GetAccessorDeclaration|SetAccessorDeclaration` the right type for member?

DevVersion replied:
Yeah this is the correct type. It just looks like TypeScript isn't exposing the `symbol` property in their typings. The same is happening on Dgeni as well ([see here](https://github.com/angular/dgeni-packages/blob/master/typescript/src/services/TsParser/index.ts#L56))

DevVersion commented:
@jelbourn Addressed your feedback.
